### DISCUSSION TITLE
[10 Giugno 2022] - Pipeline

### DISCUSSION THREAD

**Discussion by notedo** [CS-Swap/Architettura-degli-Elaboratori#25] [Pipeline]
### Traccia Esercizio

Il programma qui sotto effettua la somma dei valori assoluti (in $s0) degli elementi presenti nell’array A di lunghezza L. Se il risultato è maggiore di 10, viene stampata la stringa S: si noti che, dato l’input fornito, il risultato atteso è 16 e per ciò la stringa viene stampata.
Si supponga che tutte le istruzioni impiegate fanno parte del set supportato dalla CPU in figura, ossia non si fa uso di alcuna pseudoistruzione.
Si indichino (ignorando hazard che possano concernere la syscall):
1) le istruzioni tra le quali sono presenti data hazard – indicare i numeri di istruzione N ed M (visibili alla sinistra delle linee di codice in figura) ed il registro coinvolto $xY (nel formato N / M / $xY);
2) le istruzioni tra le quali sono presenti control hazard – indicare i numeri di istruzione N ed M (nel formato N / M);
3) quanti cicli di clock sono necessari ad eseguire il programma tramite forwarding, spiegando il calcolo effettuato;
4) quanti cicli di clock sarebbero necessari ad eseguire il programma senza forwarding, spiegando il calcolo effettuato;
5) quali sono, per ognuna delle cinque fasi, le istruzioni (o le bolle) in pipeline durante il 13° ciclo di clock (con forwarding);
6) qualora la decisione delle istruzioni di branching sia compiuta in fase EX e la policy di gestione sia di “branch not taken” by default, quante fasi di stallo causi potenzialmente ogni control hazard. Chiarire se l’uso del forwarding possa influenzare questo fenomeno o meno.

<img width="421" alt="img" src="https://github.com/CS-Swap/Architettura-degli-Elaboratori/assets/110790224/bffc9801-76b8-4781-b3f2-57f5429f4b0d">

<img width="870" alt="img" src="https://github.com/CS-Swap/Architettura-degli-Elaboratori/assets/110790224/3ec80922-f243-4c98-bf74-8b18d190a29d">

**Comment by thegeek-sys** [comment]
# Senza forwarding
<img width="993" alt="Screenshot 2024-06-03 alle 22 53 37" src="https://github.com/CS-Swap/Architettura-degli-Elaboratori/assets/55345633/a8e41b1b-e3cd-4b4f-a9c3-eb275f728ddd">

# Con forwarding
<img width="758" alt="Screenshot 2024-06-03 alle 22 54 10" src="https://github.com/CS-Swap/Architettura-degli-Elaboratori/assets/55345633/26203b7d-5d28-45c9-89f2-47f6363975cc">

# Data hazard e control hazard
<img width="258" alt="Screenshot 2024-06-03 alle 22 54 28" src="https://github.com/CS-Swap/Architettura-degli-Elaboratori/assets/55345633/819646c6-b06b-4fa7-ae5b-9eb277333cba">

# Istruzioni 13° ciclo di clock
<img width="104" alt="Screenshot 2024-06-03 alle 22 55 44" src="https://github.com/CS-Swap/Architettura-degli-Elaboratori/assets/55345633/1aec0a35-5c40-4a7d-bcd1-3fc347351057">
